File: packages/core/src/battle/FleetFactor.ts
import { FleetType, ShipRole, Side } from "../common"

type FleetFactorRule = {
  player: FleetType
  enemy: FleetType
  playerFactors: [number, number]
  enemyFactors: [number, number]
}

const shellingPowerRules: FleetFactorRule[] = [
  {
    player: "Single",
    enemy: "Combined",
    playerFactors: [5, 5],
    enemyFactors: [10, -5],
  },
  {
    player: "CarrierTaskForce",
    enemy: "Single",
    playerFactors: [2, 10],
    enemyFactors: [10, 5],
  },
  {
    player: "SurfaceTaskForce",
    enemy: "Single",
    playerFactors: [10, -5],
    enemyFactors: [5, -5],
  },
  {
    player: "TransportEscort",
    enemy: "Single",
    playerFactors: [-5, 10],
    enemyFactors: [10, 5],
  },
  {
    player: "CarrierTaskForce",
    enemy: "Combined",
    playerFactors: [2, -5],
    enemyFactors: [10, -5],
  },
  {
    player: "SurfaceTaskForce",
    enemy: "Combined",
    playerFactors: [2, -5],
    enemyFactors: [10, -5],
  },
  {
    player: "TransportEscort",
    enemy: "Combined",
    playerFactors: [-5, -5],
    enemyFactors: [10, -5],
  },
]

const shellingAccuracyRules = [
  {
    fleetType: "CarrierTaskForce",
    factors: [78, 43],
  },
  {
    fleetType: "SurfaceTaskForce",
    factors: [46, 70],
  },
  {
    fleetType: "TransportEscort",
    factors: [51, 46],
  },
]

type FleetState = {
  side: Side
  role: ShipRole
  fleetType: FleetType
}

export const getShellingFleetFactor = (attacker: FleetState, defender: FleetState) => {
  const [player, enemy] = attacker.side === "Player" ? [attacker, defender] : [defender, attacker]
  // 通常vs通常
  if (player.fleetType === "Single" && enemy.fleetType === "Single") {
    return 0
  }

  const found = shellingPowerRules.find((rule) => rule.player === player.fleetType && rule.enemy === enemy.fleetType)
  if (!found) return 0

  const factorIndex = attacker.role === "Main" ? 0 : 1
  const factors = attacker.side === "Player" ? found.playerFactors : found.enemyFactors

  return factors[factorIndex]
}

export const getTorpedoFleetFactor = (attacker: FleetState, defender: FleetState) => {
  const [player, enemy] = attacker.side === "Player" ? [attacker, defender] : [defender, attacker]

  if (enemy.fleetType !== "Single") return 15
  if (player.fleetType === "Single") return 5
  return 0
}

export const getShellingAccuracyFleetFactor = ({ fleetType, role }: FleetState) => {
  const factorIndex = role === "Main" ? 0 : 1
  const factor = shellingAccuracyRules.find((rule) => rule.fleetType === fleetType)?.factors[factorIndex]
  return factor || 90
}

export const getFleetFactors = (attacker: FleetState, defender: FleetState) => {
  const shellingPower = getShellingFleetFactor(attacker, defender)
  const shellingAccuracy = getShellingAccuracyFleetFactor(attacker)

  const torpedoPower = getTorpedoFleetFactor(attacker, defender)

  return {
    shellingPower,
    shellingAccuracy,
    torpedoPower,
  }
}

File: packages/core/src/attack/fleet_factor.rs
use crate::types::{OrgType, Role, Side};

struct FleetFactorRule {
    player: OrgType,
    enemy: OrgType,
    player_factors: (i32, i32),
    enemy_factors: (i32, i32),
}

impl FleetFactorRule {
    fn find_factor(
        &self,
        player_org_type: OrgType,
        enemy_org_type: OrgType,
        side: Side,
        role: Role,
    ) -> Option<i32> {
        (self.player == player_org_type && self.enemy == enemy_org_type).then(|| {
            let factors = if side.is_player() {
                self.player_factors
            } else {
                self.enemy_factors
            };

            if role.is_main() {
                factors.0
            } else {
                factors.1
            }
        })
    }
}

const SHELLING_POWER_RULES: [FleetFactorRule; 7] = [
    FleetFactorRule {
        player: OrgType::Single,
        enemy: OrgType::EnemyCombined,
        player_factors: (5, 5),
        enemy_factors: (10, -5),
    },
    FleetFactorRule {
        player: OrgType::CarrierTaskForce,
        enemy: OrgType::Single,
        player_factors: (2, 10),
        enemy_factors: (10, 5),
    },
    FleetFactorRule {
        player: OrgType::SurfaceTaskForce,
        enemy: OrgType::Single,
        player_factors: (10, -5),
        enemy_factors: (5, -5),
    },
    FleetFactorRule {
        player: OrgType::TransportEscort,
        enemy: OrgType::Single,
        player_factors: (-5, 10),
        enemy_factors: (10, 5),
    },
    FleetFactorRule {
        player: OrgType::CarrierTaskForce,
        enemy: OrgType::EnemyCombined,
        player_factors: (2, -5),
        enemy_factors: (10, -5),
    },
    FleetFactorRule {
        player: OrgType::SurfaceTaskForce,
        enemy: OrgType::EnemyCombined,
        player_factors: (2, -5),
        enemy_factors: (10, -5),
    },
    FleetFactorRule {
        player: OrgType::TransportEscort,
        enemy: OrgType::EnemyCombined,
        player_factors: (-5, -5),
        enemy_factors: (10, -5),
    },
];

pub fn find_shelling_power_factor(
    player_org_type: OrgType,
    enemy_org_type: OrgType,
    side: Side,
    role: Role,
) -> i32 {
    SHELLING_POWER_RULES
        .iter()
        .find_map(|rule| rule.find_factor(player_org_type, enemy_org_type, side, role))
        .unwrap_or_default()
}

pub fn find_shelling_accuracy_factor(org_type: OrgType, side: Side, role: Role) -> i32 {
    if side.is_enemy() {
        return 0;
    }

    let factors = match org_type {
        OrgType::CarrierTaskForce => (78, 43),
        OrgType::SurfaceTaskForce => (46, 70),
        OrgType::TransportEscort => (51, 46),
        _ => return 0,
    };

    if role.is_main() {
        factors.0
    } else {
        factors.1
    }
}

pub fn find_torpedo_power_factor(player_org_type: OrgType, enemy_org_type: OrgType) -> i32 {
    if enemy_org_type.is_combined() {
        15
    } else if player_org_type.is_single() {
        5
    } else {
        0
    }
}
